daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala

```diff
@@ -971,6 +1006,34 @@ final class DIArray(
   override def isSimple = false
   override def isComplex = false
 
+  // Parsers don't actually call setHidden on DIArrays, only on DIElements. But
+  // DIArrays are always created when there is at least one child element, and
+  // since all children of an array must have the same visibility, we can just
+  // inspect the isHidden property of the first child to determine if the
+  // entire array is hidden.
+  override def isHidden = _contents(0).isHidden
+
+  // Marks/blocks are never taken on arrays, so the infosetWalkerBlockCount
+  // would normally be zero for arrays. The infosetWalkerBlockCount is used to determine
+  // if this array is known to exist or if it could be backtracked. We know
+  // this DIArray exists if the first child is known to exist (i.e. if it has
+  // no blocks).
+  //
+  // Additionally, the infoset walker may delete elements from an array. In
+  // which case _contents(0) may be null. But if _contents(0) is null, that
+  // means there was an element that was known to exist, and it was freed after
+  // its event were created. So that doesn't cause a block.
+  //
+  // We also add the parent block count. This is because when we are
+  // speculatively parsing elements of an array, marks are taken before array
+  // elements are added, which increments the infosetWalkerBlockCount of the
+  // parent of this array. So if the block count of the parent is > 0, then
+  // that means we are speculatively parsing a new child, and so should not
+  // walk elements of the array.
```

Comment:
Because the block only exists while speculatively parsing a single array element. Once an array element is finished and the speculation is over, the PoU associated with that that speculation is removed, we immeidately call walk() in the SequenceBaseParser, and we can walk into this new element (assuming there are no other PoU's causing a block). 

Does this updated comment make this more clear?
```scala
  // We also add the parent block count. This is because a mark is taken when
  // speculatively parsing a single element of an array. This mark increments
  // the infosetWalkerBlockCount of the /parent/ of the DIArray because the new
  // DIElement in this DIArray does not yet exist. So as long as we are
  // speculatively parsing a single element of an array, this block on the
  // parent prevents walking into this new speculatively parsed array element.
  // Once we finish speculatively parsing that single array element, that block
  // is removed, and we immediately call walk() to walk the new array element
  // (assuming there are no other PoU's in scope). This process is repeated for
  // each speculatively parsed array element.

```

Reply:
Yes. Clearer.